src/context.rs

```diff
@@ -33,76 +34,91 @@ pub struct OutstandingRequests {
     youngest: Option<Id>,
 }
 
-pub struct Context {
-    batch_counter: BatchNumber,
-    pub batches:
-        HashMap<BatchNumber, (BatchCount, Vec<serde_json::value::Value>, ResponsesCallback)>,
+/// Groups parameters for each request.
+pub enum RequestParams<T> {
+    /// Replicates the same list of parameters for all language servers in a context.
+    All(Vec<T>),
+    /// Uses different parameters for each language server in a context.
+    Each(HashMap<ServerName, Vec<T>>),
+}
+
+pub struct ServerSettings {
+    pub root_path: String,
+    pub offset_encoding: OffsetEncoding,
+    pub preferred_offset_encoding: Option<OffsetEncoding>,
     pub capabilities: Option<ServerCapabilities>,
-    pub completion_items: Vec<CompletionItem>,
+    pub tx: Sender<ServerMessage>,
+}
+
+pub struct Context {
+    batch_count: BatchCount,
+    pub batch_sizes: HashMap<BatchNumber, HashMap<ServerName, usize>>,
+    pub batches: HashMap<
+        BatchNumber,
+        (
+            Vec<(ServerName, serde_json::value::Value)>,
+            ResponsesCallback,
+        ),
+    >,
+    pub code_lenses: HashMap<String, Vec<(ServerName, CodeLens)>>,
+    pub completion_items: Vec<(ServerName, CompletionItem)>,
     pub completion_items_timestamp: i32,
     // We currently only track one client's completion items, to simplify cleanup (else we
```

Comment:
Impressive work! A couple of comments (I didn't review the whole thing yet):

To optimzie the changes for the reader, can you structure the history so that every commit passes all tests (while being as small as practically possible).
See https://www.bitsnbites.eu/git-history-work-log-vs-recipe/
In particular, I don't care about any "fix" commits that fix bugs in your earlier commits. They should be squashed.

I'd probaly make one commit that extracts the `ServerSettings` type (and makes sure everything still compiles) without changing any behavior.
Then the actual logic changes will be much smaller.

> [language.typescriptreact]

While this may work in practice, I don't think this is kosher.
The "typescriptreact" part is sent to the language server as language ID.
The default config use "jsx" and "tsx" because those happen to work for "javascript" resp. "typescript" with the language servers I tried.

> Breaking changes

Don't. An old config file should keep working. Perhaps we want to make the parser accept both styles.

> [language.typescriptreact.tsserver]
> filetypes = ["javascript", "typescript"]
> roots = ["package.json", "tsconfig.json", "jsconfig.json", ".git", ".hg"]
> command = "typescript-language-server"
> args = ["--stdio"]
> 
> [language.typescriptreact.tailwindcss]
> filetypes = ["javascript", "typescript"]
> roots = ["tailwind.config.js", "tailwind.config.ts"]
> command = "tailwindcss-language-server"
> args = ["--stdio"]
> settings_section = "tailwindCSS"

what's wrong with

	[language.tsserver]
	filetypes = ["javascript", "typescript"]
	...
	[language.tailwindcss]
	filetypes = ["javascript", "typescript"]
	...

Reply:
> Impressive work! A couple of comments (I didn't review the whole thing yet):
> 
> To optimzie the changes for the reader, can you structure the history so that every commit passes all tests (while being as small as practically possible). See https://www.bitsnbites.eu/git-history-work-log-vs-recipe/ In particular, I don't care about any "fix" commits that fix bugs in your earlier commits. They should be squashed.

Sure thing! I just wanted to wait for feedback on some decisions, so I can work around tests based on that feedback. Once I get tests passing I'll squash everything together.

> I'd probaly make one commit that extracts the `ServerSettings` type (and makes sure everything still compiles) without changing any behavior. Then the actual logic changes will be much smaller.

This sounds good, will try it!

> > [language.typescriptreact]
> 
> While this may work in practice, I don't think this is kosher. The "typescriptreact" part is sent to the language server as language ID. The default config use "jsx" and "tsx" because those happen to work for "javascript" resp. "typescript" with the language servers I tried.

What led me to use that was the TailwindCSS server not working without a proper language ID. Then I followed the [suggested values in the specification]( https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentItem) and it started working.

> > Breaking changes
> 
> Don't. An old config file should keep working. Perhaps we want to make the parser accept both styles. (EDIT: FWIW an example of a deserializer that accepts two flavors of TOML structures is [0c10b3c](https://github.com/kak-lsp/kak-lsp/commit/0c10b3c9d072c8e3af0e217f5ee54e31a231dc62). Thankfully we never needed to merge that. Maybe we also don't need ambiguity here)
> 
> > [language.typescriptreact.tsserver]
> > filetypes = ["javascript", "typescript"]
> > roots = ["package.json", "tsconfig.json", "jsconfig.json", ".git", ".hg"]
> > command = "typescript-language-server"
> > args = ["--stdio"]
> > [language.typescriptreact.tailwindcss]
> > filetypes = ["javascript", "typescript"]
> > roots = ["tailwind.config.js", "tailwind.config.ts"]
> > command = "tailwindcss-language-server"
> > args = ["--stdio"]
> > settings_section = "tailwindCSS"
> 
> what's wrong with
> 
> ```
> [language.tsserver]
> filetypes = ["javascript", "typescript"]
> ...
> [language.tailwindcss]
> filetypes = ["javascript", "typescript"]
> ...
> ```

The problem is that we need to use a "valid" language ID, because some servers won't work without them (as mentioned, TailwindCSS only works with specific language IDs from [the specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocumentItem)).

I agree breaking a user's config is bad, and I would prefer an approach that doesn't as well. I thought we could go with something like:
```toml
[language.tsserver]
language_id = "typescriptreact"
filetypes = ["javascript", "typescript"]

[language.tailwindcss]
language_id = "typescriptreact"
filetypes = ["javascript", "typescript"]
```

In the example above, `language_id` would be optional, and if not present, the `language.<name>` would be used instead (like it currently is).

My biggest gripe with that approach is terminology, as `language` was previously attached to a language ID, and now it is attached to a language server name.

What I propose is that we have two hashmaps --- `language` and `language_server` --- and that they get merged together when config loads. This way we can avoid a breaking change by making `language` and `language_server` aliases, while also allowing the correct terminology to be used in newer versions of the config.

Here's an example of a valid config:

```toml
# This is the old config the user already has,
# and this shall not break by any means.
[language.zig]
filetypes = ["zig"]
command = "zls"

# This is the new config the user can now use.
[language_server.tsserver]
language_id = "typescriptreact"
filetypes = ["javascript", "typescript"]
command = "..."
```

Reply:
your proposal with `language_server` instead of `language` makes a ton of sense - there can be multiple servers for the same language.

Reply:
you're right, `typescriptreact` is the correct language ID. Here's an idea, maybe we want to write

```toml
[[language.typescriptreact]]
command = "typescript-language-server"
...
[[language.typescriptreact]]
command = "tailwindcss-language-server"
...
...
```

haven't tried it so not sure

Reply:
> you're right, typescriptreact is the correct language ID. Here's an idea, maybe we want to write
>
> ```toml
> [[language.typescriptreact]]
> command = "typescript-language-server"
> ...
> [[language.typescriptreact]]
> command = "tailwindcss-language-server"
> ...
> ...
> ```
> haven't tried it so not sure

This one is elegant. The only change we'd need to make is adding an optional `server` field, so we can identify each server. It needs to be optional in order to still be backwards compatible. If no server name is informed, we can use the language ID, and if it's a list, we can append an index to it (resulting in `(typescriptreact-1, typescriptreact-2)` for example). It's ugly but not the end of the world, and we can always provide examples in docs explaining that passing a `server` makes info output better. Such behavior can be achieved by using a custom deserializer, allowing the following:

```toml
# Existing config.
[language.rust]
filetypes = ["rust"]
command = "rust-analyzer"

[[language.typescriptreact]]
server = "tsserver"
filetypes = ["javascript", "typescript"]
command = "typescript-language-server"
args = ["--stdio"]

[[language.typescriptreact]]
server = "tailwindcss"
filetypes = ["javascript", "typescript"]
command = "tailwindcss-language-server"
args = ["--stdio"]
```

Reply:
@krobelus I tried implementing the approach above and it doesn't work well with TOML, specially when defining `settings` for a language server.

I think maybe the aliasing approach is simpler and easier to maintain.

Reply:
@krobelus I tried the approach above and 1) it didn't work properly and 2) is difficult to implement and hard to maintain.

Maybe we could go with the aliasing approach? It's still backwards compatible and is a lot cleaner.

Reply:
I don't see why it wouldn't work but we can handle this later. It doesn't block other work and changes to this will be small and isolated